sql-plugin/src/main/scala/com/nvidia/spark/rapids/spill/SpillFramework.scala

```diff
@@ -452,26 +500,73 @@ class SpillableDeviceBufferHandle private (
     materialized
   }
 
+  // used to gate when a spill is actively being done so that a second thread won't
+  // also begin spilling, and a handle won't release the underlying buffer if it's
+  // closed while spilling
+  private var toSpill: Option[DeviceMemoryBuffer] = None
+
+  // used to ensure the resource gets cleaned up properly after spilling and avoids
+  // double freeing
+  private var spilled: Option[DeviceMemoryBuffer] = None
+  override def releaseDeviceResource(): Unit = {
+    synchronized {
+      spilled.foreach(_.close())
+      spilled = None
+    }
+    super.releaseDeviceResource()
+  }
+
   override def spill(): Long = {
     if (!spillable) {
       0L
     } else {
-      synchronized {
-        if (host.isEmpty && dev.isDefined) {
-          host = Some(SpillableHostBufferHandle.createHostHandleFromDeviceBuff(dev.get))
-          sizeInBytes
+      val thisThreadSpills = synchronized {
+        if (host.isEmpty && dev.isDefined && toSpill.isEmpty) {
+          toSpill = dev
+          // incRefCount here so that if close() is called
+          // while we are spilling, we will prevent the buffer being freed
+          toSpill.get.incRefCount()
+          true
         } else {
-          0L
+          false
         }
       }
+      if (thisThreadSpills) {
+        withResource(toSpill.get) { buf =>
+          // the spill IO is non-blocking as it won't impact dev or host directly
+          // instead we "atomically" swap the buffers below once they are ready
+          var stagingHost: Option[SpillableHostBufferHandle] =
+            Some(SpillableHostBufferHandle.createHostHandleFromDeviceBuff(buf))
+          synchronized {
+            if (closed) {
+              stagingHost.foreach(_.close())
+              stagingHost = None
+              doClose()
+            } else {
+              host = stagingHost
+            }
+            spilled = dev
+            dev = None
+            toSpill = None
+          }
+        }
+        sizeInBytes
+      } else {
+        0
+      }
     }
   }
 
-  override def close(): Unit = {
+  private def doClose(): Unit = synchronized {
     releaseDeviceResource()
```

Comment:
so now `releaseDeviceResource` is inside of the handle lock. This method ends up calling the spill store and takes its lock. I was trying to avoid this to prevent lock inversion deadlocks. Do you have a reason to move this inside of the handle lock?

Reply:
I think it shouldn't be needed, let me try to undo that.